stablehlo/dialect/StablehloTypeInference.h

```diff
@@ -0,0 +1,162 @@
+/* Copyright 2022 The StableHLO Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef STABLEHLO_DIALECT_STABLEHLO_TYPE_INFERENCE_H
+#define STABLEHLO_DIALECT_STABLEHLO_TYPE_INFERENCE_H
+
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/ADT/StringRef.h"
+#include "mlir/Dialect/Quant/QuantTypes.h"
+#include "mlir/Dialect/Shape/IR/Shape.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Dialect.h"
+#include "mlir/IR/DialectImplementation.h"
+#include "mlir/IR/Location.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/OpDefinition.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/TensorEncoding.h"
+#include "mlir/IR/TypeUtilities.h"
+#include "mlir/IR/Types.h"
+#include "mlir/Interfaces/InferTypeOpInterface.h"
+#include "mlir/Interfaces/SideEffectInterfaces.h"
+#include "mlir/Support/LogicalResult.h"
+#include "stablehlo/dialect/Base.h"
```

Comment:
Including stablehlo files may indicate an issue in shared code. We want these functions to be built on upstream shared machinery only, or self sufficient (like `WindowDimension`).

Reply:
Removed imports of `StablehloEnums` and `StablehloAttrs `

Another problem resolved is that in `inferWhileOp` originally there is:
```
auto condReturnTypes = cast<ReturnOp>(cond.front().back()).getOperandTypes();
```
But I can't import `StablehloOps.h` either. I tried to use a parent class `like  Op or Operator` but no success. Instead, I choose to pass in `condReturnTypes` and `bodyReturnTypes` as well. Please let me know if there's a better way.

Reply:
`cond.front().back()` is an `Operation` which declares [`getOperandTypes`](https://github.com/llvm/llvm-project/blob/d376b2667a559bf8af5dc43f1a118eb10538cea5/mlir/include/mlir/IR/Operation.h#L314). Can you give it a try without a cast?

Reply:
Thanks that works without even `cast< Operation>`. Fixed.